File: Algorithms/1465_Maximum_Area_of_a_Piece_of_Cake_After_Horizontal_and_Vertical_Cuts/Java/MaximumAreaofaPieceofCakeAfterHorizontalandVerticalCutsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maxArea(int h, int w, int[] horizontalCuts, int[] verticalCuts) {
        Arrays.sort(horizontalCuts);
        Arrays.sort(verticalCuts);

        int hMax = horizontalCuts[0], vMax = verticalCuts[0];

        for (int i = 1; i < horizontalCuts.length; i++) {
            hMax = Math.max(hMax, horizontalCuts[i] - horizontalCuts[i - 1]);
        }
        hMax = Math.max(hMax, h - horizontalCuts[horizontalCuts.length - 1]);

        for (int i = 1; i < verticalCuts.length; i++) {
            vMax = Math.max(vMax, verticalCuts[i] - verticalCuts[i - 1]);
        }
        vMax = Math.max(vMax, w - verticalCuts[verticalCuts.length - 1]);

        return (int) (vMax * 1L * hMax % (Math.pow(10, 9) + 7));
    }
}

File: Algorithms/1465_Maximum_Area_of_a_Piece_of_Cake_After_Horizontal_and_Vertical_Cuts/Python/Maximum_Area_of_a_Piece_of_Cake_After_Horizontal_and_Vertical_Cuts_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def maxArea(self, h, w, horizontalCuts, verticalCuts):
        horizontalCuts = [0] + horizontalCuts + [h]
        verticalCuts = [0] + verticalCuts + [w]

        horizontalCuts = sorted(horizontalCuts)
        verticalCuts = sorted(verticalCuts)

        h_max = max(horizontalCuts[i] - horizontalCuts[i - 1] for i in range(1, len(horizontalCuts)))
        v_max = max(verticalCuts[i] - verticalCuts[i - 1] for i in range(1, len(verticalCuts)))

        return (h_max * v_max) % (pow(10, 9) + 7)





